Add requirements document for Java SPI support in the framework (#924)
osgi/osgi: .design/requirements-372.md

Multiple JDBC driver bundles are installed, each containing a `META-INF/services/java.sql.Driver` file.
The application bundle should discover all installed JDBC drivers without OSGi-specific metadata in any bundle.

### UC2: Logging Backend Discovery

A library bundle uses SLF4J, which internally calls `ServiceLoader.load(SLF4JServiceProvider.class)`.
A logging backend bundle (e.g., Logback) is installed with a standard `META-INF/services/` declaration.
SLF4J should discover the logging backend through normal `ServiceLoader` usage.

### UC3: Standard Java XML Processing

An application calls `DocumentBuilderFactory.newInstance()`, which internally uses `ServiceLoader` to discover XML parser implementations.
The framework should provide visibility to XML parser providers in other bundles that export the relevant packages.

### UC4: Mixed Environment

Some provider bundles have OSGi-specific `osgi.serviceloader` capabilities (supporting the Mediator spec) while others only have standard `META-INF/services/` files.
Both styles of provider should be discoverable.
The core SPI mechanism and the Service Loader Mediator should coexist without conflicts.

> 

Reviewer: This is an important point in my opinion. Because bundles with Service Loader Mediator  headers will probably be in use for a long time.
But besides the cases where consumer and provider are both using Service Loader Mediator or both are not using it, we also consider and support the case where one of both uses the Mediator.
This could be resolved by assuming that the required headers/wires to 'just' work as expected exists for the participant that's not using the Mediator (actively).

Reviewer: > The core SPI mechanism and the Service Loader Mediator should coexist without conflicts.

Should this section be enhanced to state that the both mechanisms should also work together? With just coexistence we would have separate world, which IMO wouldn't be good.

Author: > Should this section be enhanced to state that the both mechanisms should also work together? With just coexistence we would have separate world, which IMO wouldn't be good.

I don't understand what the difference is between "coexist without conflicts" and "work together". At least I do not want to imply any coupling between the both. The core support has to leave alone bundles that use the mediator already (as the should not change) so it just needs to detect the declaration of Mediator Requirements as a bare minimum. This does not mean a framework decides to simply implement *both* at the same time... but this is nothing the specification has to decide.

Reviewer: Yes, in general there shouldn't be a coupling between both.

> This does not mean a framework decides to simply implement both at the same time

But for the case when both is implemented I'd recommend some kind of interoperability in order to a smooth transition for 'users' of the mediator spec to the new core requirement.
Because otherwise a provider decides to rely on the new core spec and removes all mediator headers, all consumers would have to be adapted too.

And likewise a consumer that doesn't have the mediator headers will continue to not function if the provider still has the mediator headers.
So I'd add at least the recommendation that, if the Mediator spec is also implemented, both should work together. And yes that would add some coupling.

Alternatively the new core spec could mandate that the mediator spec must not be implemented/adhered at the same, which means effectively all headers from the medaitor spec would have to be ignored.
This would probably be the simpler way to ensure a smooth transition.
I assume that usually the serviceloader headers publish all advertised provided services any ways, so there shouldn't be a difference.

Only specific attribute filters in consumers would be ignored. But I again assume that they are not widely used since it only works in the OSGi world and not in plain Java. And the `ServiceLoader` API suggests other strategies to select a suitable implementation any ways.

### UC5: Module Boundary Enforcement

A consumer bundle imports package `com.example.api` which contains a Service Type.
Multiple provider bundles also import `com.example.api` and contain Provider Configuration Files.
A separate bundle that does *not* import `com.example.api` should *not* have its provider configuration files discovered by the consumer.

> The framework must respect wiring boundaries.

Reviewer: That's an important point to consider to prevent class-loading/casting issues, also especially with multiple versions of a `...api` bundle available in the runtime.
But that should be feasible since the Framework can simply derive the package name of the Service and can then determine the exporter of that Service interface/class for the currently consuming bundle.

### UC6: Dynamic Bundle Lifecycle

Provider bundles may be installed, started, stopped, and uninstalled at runtime.
A subsequent `ServiceLoader.load()` call should reflect the current set of resolved and wired provider bundles.

> Previously cached `ServiceLoader` instances are not required to update dynamically, but fresh calls must reflect current state.

Reviewer: Given that ServiceLoader itself is closed, I assume we cannot do much but consider the bundle state at the time the ServiceLoader reads the `Provider Configuration File` or attempts to load a ServiceProvider class. The latter two actions are the only ones I'm aware of where the ServiceLoader interacts with the Framework and where we can do anything.
E.g. if a Service Provider is loaded from an already uninstalled/stopped bundle, a corresponding Exception could/should be thrown so that the ServiceLoader tries another provider or fails accordingly.

> Previously cached `ServiceLoader` instances are not required to update dynamically, but fresh calls must reflect current state.

Author: > What should happen to the ServiceLoader's iterator?

As pointed out by @HannesWell we do not can (or should) care about the internal structure of the loader and must assume that an SPI is used in a consistent way during a period of time where the system state is stable.

So we can not enforce any dynamics here, if at the time of the call a class is not loadable it will simply throw an error (and Java SPi then does not possibly offer the class through its iterator). Weather or not the iterator is lazy depend on the JVM implementation but one would assume it is.

### UC7: Fragment Bundles

A fragment bundle attached to a host bundle contains a Provider Configuration File.
The framework should treat the fragment's resources as part of the host bundle's content and include them in SPI discovery when the host is wired appropriately.

## Requirements

### R1: Transparent SPI Resource Discovery

The framework must enable `ServiceLoader` to discover Provider Configuration Files from provider bundles when the consumer bundle and the provider bundle are both wired to the same Service Type package.
This must work without requiring any OSGi-specific metadata beyond standard `Import-Package` / `Export-Package` declarations.

### R2: Cross-Bundle Class Loading for SPI Implementations

When `ServiceLoader` attempts to instantiate a Service Provider class declared in a Provider Configuration File from another bundle, the framework must enable loading of that class from the provider bundle that declared it.

> This must work even though the consumer bundle does not explicitly import the package containing the Service Provider class or it is even not exported at all.

Reviewer: ```suggestion
This must work even [if] [...] the package containing the Service Provider class [...] is even not exported at all.
```
If or if not this should be the case, could be a philosophical question.
I think we should also consider if it's possible to load the provider using `Bundle.loadClass()` to keep implementations simpler?

Author: > I think we should also consider if it's possible to load the provider using Bundle.loadClass() to keep implementations simpler?

This is effectively the same... just that `Bundle.loadClass()` is a technical detail and I don't want to dictate here how a framework decides it can load the class at all.
